core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -212,6 +225,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float lhs_tmp = __half2float(val),
+                     rhs_tmp = __half2float(src.val);
+    volatile float* lhs      = &lhs_tmp;
+    volatile float* rhs      = &rhs_tmp;
+    val                      = __float2half(lhs[0] + rhs[0]);
+#else
+    // Cuda 10 supports __half volatile stores but not volatile arithmetic
+    // operands so we use atomicAdd for cuda 10.0 instead.
+    atomicAdd(const_cast<impl_type*>(&val),
+              const_cast<const impl_type&>(src.val));
+#endif  // CUDA_VERSION
+#else
+    // 32bit arithmetic on host followed by 16bit atomic store.
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        __float2half(__half2float(const_cast<impl_type&>(val)) +
+                     __half2float(const_cast<const impl_type&>(src.val)));
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  void atomic_mul(const volatile half_t src) const volatile {
```

Comment:
either copy the content to where you use it (it looks like its only called once?) or rename to something like impl_volatile_mult.